# Group 11 - Project #2 
Julia Petrillo, Daniel Lesh, Robert Fisk, and Bradley Wiseman 

## **Linear Optimization and Sensitivity Analysis**
> Goal: *Maximize Profit on Selling Different Types of Oil Grades*

### **Assigned Variables by Brand of Oil** 


> *Regular Brands*:
* C1R = Crude Stock #1 blended with Regular Brand
* C2R = Crude Stock #2 blended with Regular Brand
* C3R = Crude Stock #3 blended with Regular Brand
* C4R = Crude Stock #4 blended with Regular Brand

> *Multigrade Brands*:
* C1M = Crude Stock #1 blended with Multigrade Brand
* C2M = Crude Stock #2 blended with Multigrade Brand
* C3M = Crude Stock #3 blended with Multigrade Brand
* C4M = Crude Stock #4 blended with Multigrade Brand

> *Supreme Brands*:
* C1S = Crude Stock #1 blended with Supreme Brand
* C2S = Crude Stock #2 blended with Supreme Brand
* C3S = Crude Stock #3 blended with Supreme Brand
* C4S = Crude Stock #4 blended with Supreme Brand

### **Objective Function**

> **Maximum Profit = Revenue - Cost**
* Daily Revenues = 
$8.50(C1R + C2R +C3R + C4R) + 9.00(C1M + C2M +C3M + C4M) + 10.00(C1S + C2S +C3S + C4S)$
* Daily Costs = 
$7.10(C1R + C1M + C1S) + 8.50(C2R + C2M +C 2S) + 7.70(C3R + C3M + C3S) + 9.00(C4R + C4M + C4S)$
 

### **Constraints (*Subject To*)**



> *Viscosity*:
* Regular Brand:
$(20C1R + 40C2R + 30C3R + 55C4R)/ (C1R +C2R + C3R + C4R) \geq 25$
  * Simplified: $-5C1R + 15C2R + 5C3R + 30C4R \geq 0$
* Midgrade Brands:
$(20C1M + 40C2M + 30C3M + 55C4M)/ (C1M +C2M + C3M + C4M) \geq 35$
  * Simplified: $-15C1M + 5C2M - 5C3M + 15C4M \geq 0$
* Supreme Brands:
$(20C1S + 40C2S + 30C3S + 55C4S)/ (C1S +C2S + C3S + C4S) \geq  50$
  * Simplified: $-30C1S - 10C2S - 20C3S + 5C4S \geq  0$

> *Demand*: 
* $C1R + C2R + C3R + C4R = 2000$
* $C1M + C2M + C3M + C4M = 1500$
* $C1S + C2S + C3S + C4S = 750$

> *Supply*:
* $C1R + C1M + C1S \leq 1000$
* $C2R + C2M + C2S \leq 1100$
* $C3R + C3M + C3S \leq 1200$
* $C4R + C4M + C4S \leq 1100$

> Nonnegativity:
* $C1R,C1M,C1S,C2R,C2M,C2S,C3R,C3M,C3S,C4R,C4M,C4S \in \mathbb{R}^+$











## **Importing Personal Drive and Module to Solve Linear Optimization**




In [ ]:
# Mount personal Google Drive using folder on the left
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [1]:
# Insert program to perform Linear Optimization using the Pyomo Package
%matplotlib inline
from pylab import *

import shutil
import sys
import os.path

if not shutil.which("pyomo"):
    !pip install -q pyomo
    assert(shutil.which("pyomo"))

if not (shutil.which("glpsol") or os.path.isfile("glpsol")):
    if "google.colab" in sys.modules:
        !apt-get install -y -qq glpk-utils
    else:
        try:
            !conda install -c conda-forge ipopt
        except:
            pass

assert(shutil.which("glpsol") or os.path.isfile("glpsol"))

from pyomo.environ import *

SOLVER = 'glpk'
EXECUTABLE = '/usr/bin/glpsol'

     |████████████████████████████████| 9.5MB 5.2MB/s 
     |████████████████████████████████| 51kB 5.9MB/s 
     |████████████████████████████████| 256kB 50.4MB/s 
     |████████████████████████████████| 163kB 46.0MB/s 
Selecting previously unselected package libsuitesparseconfig5:amd64.
(Reading database ... 160983 files and directories currently installed.)
Preparing to unpack .../libsuitesparseconfig5_1%3a5.1.2-2_amd64.deb ...
Unpacking libsuitesparseconfig5:amd64 (1:5.1.2-2) ...
Selecting previously unselected package libamd2:amd64.
Preparing to unpack .../libamd2_1%3a5.1.2-2_amd64.deb ...
Unpacking libamd2:amd64 (1:5.1.2-2) ...
Selecting previously unselected package libcolamd2:amd64.
Preparing to unpack .../libcolamd2_1%3a5.1.2-2_amd64.deb ...
Unpacking libcolamd2:amd64 (1:5.1.2-2) ...
Selecting previously unselected package libglpk40:amd64.
Preparing to unpack .../libglpk40_4.65-1_amd64.deb ...
Unpacking libglpk40:amd64 (4.65-1) ...
Selecting previously unselected package glpk-

## **Create Concrete Model and Declare Variables**

In [12]:
# Declare the Model using ConcreteModel() function 
model = ConcreteModel()

### Need to tell Pyomo Package which variables our model has
  ##  Use function called Var() and assign its return value to an object of the model 
    # Assuming we are using any nonnegative real value (i.e. a positive number)
# Crude Stock 1 blended with Regular, Multigrade, and Supreme Oil 
model.C1R = Var(domain=NonNegativeReals)
model.C1M = Var(domain=NonNegativeReals)
model.C1S = Var(domain=NonNegativeReals)

# Crude Stock 2 blended with Regular, Multigrade, and Supreme Oil 
model.C2R = Var(domain=NonNegativeReals)
model.C2M = Var(domain=NonNegativeReals)
model.C2S = Var(domain=NonNegativeReals)

# Crude Stock 3 blended with Regular, Multigrade, and Supreme Oil 
model.C3R = Var(domain=NonNegativeReals)
model.C3M = Var(domain=NonNegativeReals)
model.C3S = Var(domain=NonNegativeReals)

# Crude Stock 4 blended with Regular, Multigrade, and Supreme Oil 
model.C4R = Var(domain=NonNegativeReals)
model.C4M = Var(domain=NonNegativeReals)
model.C4S = Var(domain=NonNegativeReals)

## **Objective Function and Constraints**


In [13]:
## Establish the objective of our model 
  # Objective is to MAXIMIZE Profit 
model.profit = Objective(
                      expr = 8.5*model.C1R + 8.5*model.C2R + 8.5*model.C3R + 
                      8.5*model.C4R + 9.00*model.C1M + 9.00*model.C2M + 
                      9.00*model.C3M + 9.00*model.C4M + 10.00*model.C1S + 
                      10.00*model.C2S + 10.00*model.C3S + 10.00*model.C4S - 
                      7.10*model.C1R - 7.10*model.C1M - 7.10*model.C1S - 
                      8.50*model.C2R - 8.50*model.C2M - 8.50*model.C2S - 
                      7.70*model.C3R - 7.70*model.C3M - 7.70*model.C3S - 
                      9.00*model.C4R - 9.00*model.C4M - 9.00*model.C4S,
                      sense = maximize # 'Sense' indicates whether the problem is minimization or maximization <-- In this problem, we are Maximizing
                        )

## Constraints that are Subject to Viscosity
  # Must use >= to represent the inequality for these set constraints
model.RegularBrand = Constraint(expr = (-5*model.C1R) + (15*model.C2R) + (5*model.C3R) + (30*model.C4R) >= 0)
model.MultigradeBrand = Constraint(expr = (-15*model.C1M) + (5*model.C2M) - (5*model.C3M) + (15*model.C4M) >= 0)
model.SupremeBrand = Constraint(expr = (-30*model.C1S) - (10*model.C2S) - (20*model.C3S) + (5*model.C4S) >= 0)

## Constraints that are Subject to Demand
  # Must use == to represent the equality for these set constraints
model.RegularDemand = Constraint(expr = model.C1R + model.C2R + model.C3R + model.C4R == 2000)
model.MultigradeDemand = Constraint(expr = model.C1M + model.C2M + model.C3M + model.C4M == 1500)
model.SupremeDemand = Constraint(expr = model.C1S + model.C2S + model.C3S + model.C4S == 750)

## Constraints Subject to Supply
  # Must use <= to represent the inequality for these set constraints
model.CrudeStock1Supply = Constraint(expr = model.C1R+model.C1M+model.C1S <= 1000)
model.CrudeStock2Supply = Constraint(expr = model.C2R+model.C2M+model.C2S <= 1100)
model.CrudeStock3Supply = Constraint(expr = model.C3R+model.C3M+model.C3S <= 1200)
model.CrudeStock4Supply = Constraint(expr = model.C4R+model.C4M+model.C4S <= 1100)

# Print the Model! 
model.pprint()

12 Var Declarations
    C1M : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :     0 :  None :  None : False :  True : NonNegativeReals
    C1R : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :     0 :  None :  None : False :  True : NonNegativeReals
    C1S : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :     0 :  None :  None : False :  True : NonNegativeReals
    C2M : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :     0 :  None :  None : False :  True : NonNegativeReals
    C2R : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :     0 :  None :  None : False :  True : NonNegativeReals
    C2S : Size=1, Index=None
        Key  : Lower : Value : Upper : Fixed : Stale : Domain
        None :     0 :  None :  None : False :  True : NonNegativeReals
  

## **Solve Objective Function**

In [14]:
# Solve Model with Solver Command --> this is similar to the same function our Excel model performs when utilizing the Solver Plug-In 
SolverFactory(SOLVER, executable=EXECUTABLE).solve(model).write()

# Show the results by printing names along with values produced from the model 
print("Maximum Profit = ", model.profit(), "per day") # 'Maximum Profit' because this is a maximization problem
print("Regular Brand = ", model.RegularDemand(), "units per day")
print("Multigrade Brand = ", model.MultigradeDemand(), "units per day")
print("Supreme Brand = ", model.SupremeDemand(), "units per day")
print("Crude Stock 1 = ", model.CrudeStock1Supply(), "units per day")
print("Crude Stock 2 = ", model.CrudeStock2Supply(), "units per day")
print("Crude Stock 3 = ", model.CrudeStock3Supply(), "units per day")
print("Crude Stock 4 = ", model.CrudeStock4Supply(), "units per day")

# ==========================================================
# = Solver Results                                         =
# ==========================================================
# ----------------------------------------------------------
#   Problem Information
# ----------------------------------------------------------
Problem: 
- Name: unknown
  Lower bound: 3760.0
  Upper bound: 3760.0
  Number of objectives: 1
  Number of constraints: 11
  Number of variables: 13
  Number of nonzeros: 37
  Sense: maximize
# ----------------------------------------------------------
#   Solver Information
# ----------------------------------------------------------
Solver: 
- Status: ok
  Termination condition: optimal
  Statistics: 
    Branch and bound: 
      Number of bounded subproblems: 0
      Number of created subproblems: 0
  Error rc: 0
  Time: 0.016382217407226562
# ----------------------------------------------------------
#   Solution Information
# -------------------------------

> #### **Solver Results**

* Maxmimum Profit = $3,760 per day
* Regular Brand Oil = 2,000 units per day
* Multigrade Brand Oil = 1,499.99 units per day 
* Supreme Brand Oil = 750 units per day 
* Crude Stock 1 = 1,000 units per day 
* Crude Stock 2 = 1,099.99 units per day 
* Crude Stock 3 = 1,200 units per day 
* Crude Stock 4 = 950 units per day 



## **Perform Sensitivity Analysis**

In [15]:
# Save the LP Model that was built
model.write("/content/model.lp", io_options={'symbolic_solver_labels': True}) # model.lp file will run after executing line of code 

# Generate the file "sensit.sen", which contains the report 
!/usr/bin/glpsol -m /content/model.lp --lp --ranges sensit.sen # Will use "sensit.sen" to run report built for sensitivity analysis 

#Running code below to see the contents of the sensitivity analysis report
!cat /content/sensit.sen

GLPSOL: GLPK LP/MIP Solver, v4.65
Parameter(s) specified in the command line:
 -m /content/model.lp --lp --ranges sensit.sen
Reading problem data from '/content/model.lp'...
11 rows, 13 columns, 37 non-zeros
100 lines were read
GLPK Simplex Optimizer, v4.65
11 rows, 13 columns, 37 non-zeros
Preprocessing...
10 rows, 12 columns, 36 non-zeros
Scaling...
 A: min|aij| =  1.000e+00  max|aij| =  3.000e+01  ratio =  3.000e+01
GM: min|aij| =  6.389e-01  max|aij| =  1.565e+00  ratio =  2.449e+00
EQ: min|aij| =  4.082e-01  max|aij| =  1.000e+00  ratio =  2.449e+00
Constructing initial basis...
Size of triangular part is 10
      0: obj =   7.500000000e+02 inf =   1.419e+03 (2)
      3: obj =   3.520000000e+03 inf =   0.000e+00 (0)
*     6: obj =   3.760000000e+03 inf =   0.000e+00 (0)
OPTIMAL LP SOLUTION FOUND
Time used:   0.0 secs
Memory used: 0.0 Mb (43461 bytes)
Write sensitivity analysis report to 'sensit.sen'...
GLPK 4.65 - SENSITIVITY ANALYSIS REPORT                                        

In [17]:
# Obtain dual solutions from first solve and send to warm start
model.dual = Suffix(direction=Suffix.IMPORT_EXPORT)

# Obtain dual solutions from first solve and send to warm start
model.rc = Suffix(direction=Suffix.IMPORT_EXPORT)

# Obtain dual solutions from first solve and send to warm start
model.slack = Suffix(direction=Suffix.IMPORT_EXPORT)

# Solve the problem and show the results
SolverFactory(SOLVER,executable=EXECUTABLE).solve(model).write()
print("Maximum Profit = ", model.profit(), "per day") # 'Maximum Profit' because this is a maximization problem
print("Regular Brand = ", model.RegularDemand(), "units per day")
print("Multigrade Brand = ", model.MultigradeDemand(), "units per day")
print("Supreme Brand = ", model.SupremeDemand(), "units per day")
print("Crude Stock 1 = ", model.CrudeStock1Supply(), "units per day")
print("Crude Stock 2 = ", model.CrudeStock2Supply(), "units per day")
print("Crude Stock 3 = ", model.CrudeStock3Supply(), "units per day")
print("Crude Stock 4 = ", model.CrudeStock4Supply(), "units per day")

# ==========================================================
# = Solver Results                                         =
# ==========================================================
# ----------------------------------------------------------
#   Problem Information
# ----------------------------------------------------------
Problem: 
- Name: unknown
  Lower bound: 3760.0
  Upper bound: 3760.0
  Number of objectives: 1
  Number of constraints: 11
  Number of variables: 13
  Number of nonzeros: 37
  Sense: maximize
# ----------------------------------------------------------
#   Solver Information
# ----------------------------------------------------------
Solver: 
- Status: ok
  Termination condition: optimal
  Statistics: 
    Branch and bound: 
      Number of bounded subproblems: 0
      Number of created subproblems: 0
  Error rc: 0
  Time: 0.0130462646484375
# ----------------------------------------------------------
#   Solution Information
# ---------------------------------

In [18]:
print('\nShadow Prices')

# Print("Constraint ",model.demand())
print('Regular Brand:')
print('\tShadow Price = ', model.dual[model.RegularDemand]) # Displays shadow price
print('\tUpper slack = ', model.RegularDemand.uslack() ) # Displays upper slack
print('\tLower slack = ', model.RegularDemand.lslack()) # Displays lower slack

print('Multigrade Brand:')
print('\tShadow Price = ', model.dual[model.MultigradeDemand]) # Displays shadow price
print('\tUpper slack = ', model.MultigradeDemand.uslack() ) # Displays upper slack
print('\tLower slack = ', model.MultigradeDemand.lslack()) # Displays lower slack

print('Supreme Brand:')
print('\tShadow Price = ', model.dual[model.SupremeDemand]) # Displays shadow price
print('\tUpper slack = ', model.SupremeDemand.uslack() ) # Displays upper slack
print('\tLower slack = ', model.SupremeDemand.lslack()) # Displays lower slack


Shadow Prices
Regular Brand:
	Shadow Price =  -0.5
	Upper slack =  0.0
	Lower slack =  0.0
Multigrade Brand:
	Shadow Price =  0.0
	Upper slack =  2.2737367544323206e-13
	Lower slack =  -2.2737367544323206e-13
Supreme Brand:
	Shadow Price =  1.0
	Upper slack =  0.0
	Lower slack =  0.0


In [19]:
print('\nShadow Prices')

# Print("Constraint ",model.supply())
print('Crude Stock #1 Supply:')
print('\tShadow Price = ', model.dual[model.CrudeStock1Supply]) # Displays shadow price
print('\tUpper slack = ', model.CrudeStock1Supply.uslack() ) # Displays shadow price
print('\tLower slack = ', model.CrudeStock1Supply.lslack()) # Displays shadow price

print('Crude Stock #2 Supply:')
print('\tShadow Price = ', model.dual[model.CrudeStock2Supply]) # Displays shadow price
print('\tUpper slack = ', model.CrudeStock2Supply.uslack() ) # Displays shadow price
print('\tLower slack = ', model.CrudeStock2Supply.lslack()) # Displays shadow price

print('Crude Stock #3 Supply:')
print('\tShadow Price = ', model.dual[model.CrudeStock3Supply]) # Displays shadow price
print('\tUpper slack = ', model.CrudeStock3Supply.uslack() ) # Displays shadow price
print('\tLower slack = ', model.CrudeStock3Supply.lslack()) # Displays shadow price

print('Crude Stock #4 Supply:')
print('\tShadow Price = ', model.dual[model.CrudeStock4Supply]) # Displays shadow price
print('\tUpper slack = ', model.CrudeStock4Supply.uslack() ) # Displays shadow price
print('\tLower slack = ', model.CrudeStock4Supply.lslack()) # Displays shadow price



Shadow Prices
Crude Stock #1 Supply:
	Shadow Price =  1.9
	Upper slack =  0.0
	Lower slack =  inf
Crude Stock #2 Supply:
	Shadow Price =  0.5
	Upper slack =  2.2737367544323206e-13
	Lower slack =  inf
Crude Stock #3 Supply:
	Shadow Price =  1.3
	Upper slack =  0.0
	Lower slack =  inf
Crude Stock #4 Supply:
	Shadow Price =  0.0
	Upper slack =  150.0
	Lower slack =  inf
